android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt

```diff
@@ -535,37 +595,125 @@ data class ContextMenuCandidate(
         ) {
-            val clipboardManager =
-                context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
             val clip = ClipData.newPlainText(label, plainText)
-            clipboardManager.setPrimaryClip(clip)
+            context.getClipboardManager().setPrimaryClip(clip)
 
             snackbarDelegate.show(
-                snackBarParentView = snackBarParentView,
+                snackbarParentView = snackbarParentView,
                 text = displayTextId,
                 duration = Snackbar.LENGTH_SHORT,
             )
         }
+
+        @OptIn(DelicateCoroutinesApi::class)
+        @Suppress("LongParameterList")
+        private fun clipImage(
+            context: Context,
+            dispatcher: CoroutineDispatcher,
+            imageSrc: String,
+            displayTextId: Int,
+            snackbarParentView: View,
+            snackbarDelegate: SnackbarDelegate = DefaultSnackbarDelegate(),
+            systemTimeProvider: () -> Long,
+            uriForFileProvider: (Context, File) -> Uri,
+        ) {
+            val imageUrl = try {
+                URL(imageSrc)
+            } catch (e: MalformedURLException) {
+                Log.log(Log.Priority.ERROR, ContextMenuCandidate::class.java.simpleName, e)
+                return
+            }
+
+            GlobalScope.launch(dispatcher) {
+                // Concatenate the current time (milliseconds) and the URL hash code (absolute value)
+                // to generate a file name.
+                val localTempFileName = context.getString(
+                    R.string.mozac_feature_contextmenu_copy_image_filename,
+                    systemTimeProvider.invoke().toString(),
+                    imageSrc.hashCode().absoluteValue.toString(),
+                )
+
+                val localTempFile = File(context.externalCacheDir, localTempFileName)
+
+                // In the extremely unlikely event that a file with the same name already exists in
+                // the cache, overwrite it.
+                if (imageUrl.maybeCopyUrlImageTo(localTempFile)) {
+                    val contentUri = uriForFileProvider.invoke(context, localTempFile)
+                    val clipData =
+                        ClipData.newUri(context.contentResolver, "Copied image URI", contentUri)
+                    context.getClipboardManager().setPrimaryClip(clipData)
+
+                    showSnackbarConfirmation(snackbarDelegate, snackbarParentView, displayTextId)
+                }
+            }
+        }
+
+        /**
+         * Try to copy the URL content to the given [outputFile].
+         * @return true if copy was successful, otherwise false.
+         */
+        private fun URL.maybeCopyUrlImageTo(outputFile: File) = try {
+            openStream().use { inputStream ->
+                outputFile.outputStream().use { outputStream ->
+                    inputStream.copyTo(outputStream)
```

Comment:
There is an interesting 'bug' that is caught here that can be triggered if the device date\time is modified, it can trigger an SSLHandshakeException. I would like to investigate this in a separate PR for this as to not block this feature.